### PR TITLE
Fixed #32450 -- Fixed crash when ANDing/ORing an empty Q() with not pickleable Q().

### PR DESCRIPTION
**Pull request by starryrbs**
[ticket-32450](https://code.djangoproject.com/ticket/32450)

### CODE REVIEW

File: django/db/models/query_utils.py [hunk 1]
```diff
@@ -46,10 +45,10 @@ def _combine(self, other, conn):
 
         # If the other Q() is empty, ignore it and just use `self`.
         if not other:
-            return copy.deepcopy(self)
+            return self
         # Or if this Q is empty, ignore it and just use `other`.
         elif not self:
-            return copy.deepcopy(other)
+            return other
```
Comment (charettes) [inline django/db/models/query_utils.py:51]: That won't work, `_combine` must return a _new_ object.

I suggest we adapt the code to catch `TypeError` on `deepcopy` and use the old code path on failure.

Comment (starryrbs) [reply django/db/models/query_utils.py]: Thanks for your review, I think if catching the exception and using the old code does not meet our needs, can I use deconstruct to recreate an object?

**Comment by charettes** [comment]
> Thanks for your review, I think if catching the exception and using the old code does not meet our needs, can I use deconstruct to recreate an object?

`deconstruct` would certainly work for the original use case of https://github.com/django/django/pull/8517 but could you elaborate on your _needs_ as this ticket was accepted on the basis that this was causing errors on pickling and if there's something else that needs to be accounted for we might want to capture it during tests?

**Comment by starryrbs** [comment]
> > Thanks for your review, I think if catching the exception and using the old code does not meet our needs, can I use deconstruct to recreate an object?
> 
> `deconstruct` would certainly work for the original use case of #8517 but could you elaborate on your _needs_ as this ticket was accepted on the basis that this was causing errors on pickling and if there's something else that needs to be accounted for we might want to capture it during tests?

**Comment by starryrbs** [comment]
> > Thanks for your review, I think if catching the exception and using the old code does not meet our needs, can I use deconstruct to recreate an object?
> 
> `deconstruct` would certainly work for the original use case of #8517 but could you elaborate on your _needs_ as this ticket was accepted on the basis that this was causing errors on pickling and if there's something else that needs to be accounted for we might want to capture it during tests?

If the exception is caught, like this:
```
    def _combine(self, other, conn):
        if not isinstance(other, Q):
            raise TypeError(other)

        try:
            # If the other Q() is empty, ignore it and just use `self`.
            if not other:
                return copy.deepcopy(self)
            # Or if this Q is empty, ignore it and just use `other`.
            elif not self:
                return copy.deepcopy(other)
        except TypeError:
            pass

        obj = type(self)()
        obj.connector = conn
        obj.add(self, conn)
        obj.add(other, conn)
        return obj
```

There is no way to pass the following test:

```
q = Q(x__in={}.keys())
self.assertEqual(q | Q(), q)
self.assertEqual(Q() | q, q)

# AssertionError: <Q: (OR: ('x__in', dict_keys([])), (AND: ))> != <Q: (AND: ('x__in', dict_keys([])))>
```

Then the results of objects that can be deeply copied and objects that cannot be deeply copied are different, use deconstruct to solve this problem.

**Comment by charettes** [comment]
Thanks for the extra details, so you'd like to make sure the optimization introduced in https://github.com/django/django/pull/8517 is still performed even when objects are not pickleable which is fair.

**Comment by starryrbs** [comment]
> Thanks for the extra details, so you'd like to make sure the optimization introduced in #8517 is still performed even when objects are not pickleable which is fair.

Yes, I think so.

**Comment by charettes** [comment]
Note that this approach might conflict with https://github.com/django/django/pull/14017. It might be worth adding a test that captures that `Q() & Exists()` and `Exists() & Q()` work now that we're using `deconstruct`.

**Comment by starryrbs** [comment]
> @felixxm Note that this approach might conflict with #14017. It might be worth adding a test that captures that `Q() & Exists()` and `Exists() & Q()` work now that we're using `deconstruct`.
> 
> @starryrbs your PR seems good to go, no need to make the adjustments yourself.

Okay, thanks again for your review.

**Comment by felixxm** [comment]
> @felixxm Note that this approach might conflict with #14017. It might be worth adding a test that captures that `Q() & Exists()` and `Exists() & Q()` work now that we're using `deconstruct`.

`queryset` is required for `Exists()` and `Exists(qs) & Q()`/`Exists(qs) | Q()` work even without this PR. I will try `Q() & Exists(qs)` and `Q() & Exists(qs)` with #14017 after merging this one.

**Comment by felixxm** [comment]
@starryrbs Thanks :+1: I added assertions for the `AND` operator.

**Pull request merged**